### DISCUSSION TITLE
Poetry is not a Package Manager?

### DISCUSSION THREAD

**Discussion by chhex** [python-poetry/poetry#7516] [Q&A] [answered]
Looking into Poetry , i find a really great development tool... everything a developer can wish! 
Especially neat are the management of virtual environments and the 

`poetry run `

But we develop code to be used respectively run ... obviously ....  

With pip i can do: 

`python -m pip install git+https://some/path/to/a/package.py`

This is great for in house development of light-weight cli tools and applications for which python is so great... 

I may be mistaken , but this usage  scenario is not supported by poetry? Poetry assumes that packages are installed with in the context of a "project" ... it needs a pyproject.toml  file  necessarily .... but I certainly don't to be copying around pyprojedt.toml for this  simple requirement...  so poetry is mostly geared toward development of libraries?

So for installing in a  runnable package respectively a package with entry points something else is needed ... ?  

So i am back to pip respectively setuptool again , which i wanted to avoid in the first .... maybe i am missing something? 

I could use both in parallel , but lead to (terrible) redundancy in the Meta data (pyproject.toml) since Poetry has there specific entries 
```

[tool.poetry.scripts]
....

[tool.poetry.dependencies]
python = "^3.11"
beautifulsoup4 = "4.11.1"
python-jenkins = "1.7.0"
lxml =  "4.9.1"

[tool.poetry.group.dev.dependencies]
...
```

I think my main worry is that Poetry isn't https://peps.python.org/pep-0621/ compliant.....

I am looking into https://hatch.pypa.io/ as alternative....

Happy if anybody proves me wrong

**Comment by radoering** [comment, marked as answer]
> With pip i can do:
> 
> `python -m pip install git+https://some/path/to/a/package.py`

Poetry does support git dependencies as explained in the [docs](https://python-poetry.org/docs/dependency-specification#git-dependencies).

> Poetry assumes that packages are installed with in the context of a "project" ... it needs a pyproject.toml file necessarily

If you want to manage a project via poetry than you need a pyproject.toml for your project, but you can add git dependencies that are not managed by poetry.

> so poetry is mostly geared toward development of libraries?

It's for libraries as well as for applications, it's for building sdists and wheels of your project and for dependency management. If you don't want to build, you can use Poetry only for dependency management, too.

> So for installing in a runnable package respectively a package with entry points something else is needed ... ?

No. During development Poetry installs your root project in editable mode. It installs entry points defined in the [scripts](https://python-poetry.org/docs/pyproject/#scripts) section. Sdists and wheels built by Poetry can be installed by any standard conform installer.

> I could use both in parallel

That only makes sense if you want to use Poetry only for dependency management. If you use Poetry for building than you don't need setuptools.

> I think my main worry is that Poetry isn't https://peps.python.org/pep-0621/ compliant.....

That's true, but there are plans to support PEP 621 in the future: https://github.com/python-poetry/roadmap/issues/3

**Comment by chhex** [comment]
Great thanks for your answers , very appreciated. 

I am going for now with: 

- Poetry as dependency management 
- pip for installing from git and --editable locally for testing
- venv for local testing 

Quite lightweight and as near as it get to a "standard" stack: i don't have any tool specific entries in my project toml.
Poetry does add significant value for dependency management... and for me seems there ahead of the pack...
When PEP 621 is supported by Poetry, i will likely reconsider

**Reply by finswimmer** [reply]
Hey,

I don't recommend mixing Poetry with pip installs. Both resolve dependencies different, so you might end up with unexpected dependencies in your environment. As @radoering said, Poetry supports git dependencies.

> When PEP 621 is supported by Poetry, i will likely reconsider

I'm curious about what benefits you do expect from PEP 621. 🙂

**Reply by chhex** [reply]
Hi , i am not sure if we are on the same board: I use Poetry for dependency management , not for resolution. and only for convenience. So i use it basically for maintaining the 

```
dependencies = [
  .....
]
```

section.  

> so you might end up with unexpected dependencies in your environment

that's in my own responsibility 

> As @radoering said, Poetry supports git dependencies.

But only with a pyproject.toml , which as stated is not  my usage scenario 

> I'm curious about what benefits you do expect from PEP 621.

What are standards good for? Good point. I appreciate standards as means for the user to be less dependent on a specific tool chain, specially when they are community driven. Also, as in my case as casual user,  there is a larger common denominator in terms documentation and semantics. 

All that said, i find Poetry really a cool tool. Great work.  It probably just doesn't fit my marginal usage scenario...
